kvserver: extract unreplicated SST writing in applySnapshot

tbg opened the pull request:
The engine (pebble) interactions in applySnapshot are complex and
difficult to test in isolation. By extracting these responsibilities
into a standalone function, we make it easier to unit test and further
refactor the snapshot application process.

To this end, this PR introduces a new prepareSnapshot function in
snapshot_apply_prepare.go, and moves the construction of the SSTs
(and generally all engine interactions prior to the commit) there.

Epic: CRDB-46488

Release note: None

tbg committed:
kvserver: extract unreplicated SST writing in applySnapshot

The engine (pebble) interactions in applySnapshot are complex and
difficult to test in isolation. By extracting these responsibilities
into a standalone function, we make it easier to unit test and further
refactor the snapshot application process. This is the first step in a
series of incremental changes to move all engine logic out of
applySnapshot.

This change introduces a new prepareSnapshot function in
snapshot_apply_prepare.go, which encapsulates the logic for writing the
unreplicated SST during snapshot application. The relevant code in
Replica.applySnapshot is refactored to use this new helper, with the
input/output structs designed to be extended in future steps.

Technical details:
- Introduces prepareSnapshotInput and preparedSnapshot structs.
- Adds prepareSnapshot, which performs the unreplicated SST write and
  returns the cleared span.
- applySnapshot now constructs a prepareSnapshotInput and calls
  prepareSnapshot, appending the cleared span to clearedSpans.
- No functional change; this is a pure refactor.

Epic: CRDB-46488

Release note: None

tbg committed:
kvserver: move subsumed replica disk data clearing into prepareSnapshot

This change continues the refactor of snapshot application by moving the
engine logic for clearing subsumed replica disk data into the new
prepareSnapshot function. The input/output structs are expanded to
support this, and applySnapshot now passes all required fields up-front.

Epic: CRDB-46488

Release note: None

tbg committed:
kvserver: move writeUnreplicatedSST, clearSubsumedReplicaDiskData, rewriteRaftState

This was done with Goland, i.e. purely mechanical.

tbg committed:
kvserver: simplify snapshot ingestion conditional

This change refactors the logic in applySnapshot for determining whether
to apply a snapshot as a batch or via ingestion. The new structure
initializes appliedAsWrite to true and only sets it to false if the
snapshot contains external or shared SSTs, or if the SST size exceeds
the configured threshold. The conditional for ingestion is now a single
if/else, making the code easier to follow and less error-prone.

Epic: CRDB-46488

Release note: None

tbg committed:
kvserver: remove subsumedNextReplicaID from prepareSnapshotInput

The prepareSnapshotInput struct no longer needs to carry the
subsumedNextReplicaID field, as all usages can directly reference the
mergedTombstoneReplicaID constant. This simplifies the snapshot preparation and
subsumed replica cleanup logic by eliminating an unnecessary parameter and
making the use of the tombstone value explicit.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145328)
<!-- Reviewable:end -->

pav-kv commented on pkg/kv/kvserver/replica_raftstorage.go:
nit: the name is slightly misleading because the variable is true before we ingest. Consider `ingested -> ingest` or `applyAsIngest`.

pav-kv commented on pkg/kv/kvserver/replica_raftstorage.go:
Can `inSnap.sharedSSTs` and `inSnap.externalSSTs` be non-nil but zero-length? Would that change the behaviour of `IngestAndExciseFiles`? Double-checking because before this change we were passing nils, and now we only check `len=0` and pass the slices as they are (nil or `[]{}`).

pav-kv commented on pkg/kv/kvserver/snapshot_apply_prepare.go:
```diff
@@ -43,7 +49,28 @@ func prepareSnapshot(input prepareSnapshotInput) (preparedSnapshot, error) {
+		// ranges _can't_ have new replicas.
+		spans, err := clearSubsumedReplicaDiskData(
+			input.ctx, input.st, input.todoEng, input.writeSST,
+			input.desc, input.subsumedDescs, input.subsumedNextReplicaID,
```
Upd: I see it done in the last commit, nice. Arguably, the change in the last commit could have gone earlier so that the intermediate wouldn't raise a question.

tbg replied:
Yeah, agreed. It only occurred to me later to do this, and I doubt it's worth the rebase now.

pav-kv reviewed (approved):
LGTM % nits

tbg committed:
kvserver: don't store ctx in struct

It's an anti-pattern. Missed this during self-review (I made this PR
experimenting with Cursor).

tbg committed:
kvserver: rename applySnapshot{->raftMuLocked}

tbg commented on pkg/kv/kvserver/replica_raftstorage.go:
```diff
@@ -540,36 +537,29 @@ func (r *Replica) applySnapshot(
+	clearedSpans = append(clearedSpans, prepResult.clearedSubsumedSpans...)
+
+	// Update Raft entries.
+	r.store.raftEntryCache.Drop(r.RangeID)
```
I'm actually unsure if there are any specific ordering requirements here. I was hoping that the ordering would be inconsequential since `raftMu` is held throughout this method.

However, the entry cache has to accesses that may happen under `r.mu` without `r.raftMu`, namely `termLocked` and `entriesLocked`. This leads me to believe that we should evict from the cache before removing the log (committing the SSTs). Otherwise, an interleaving call to either of the above methods (which would necessarily have to occur as a result of stepping a raft message into the RawNode or flushing the proposal buffer if it's full) could end up reading entries from the entryCache that aren't actually there any more.

Though this might all be fine (not to mention that if we're applying a snapshot, there shouldn't be any interesting entry reads that can happen, since we're behind on the log).

This all breaks my brain a little bit. I'm going to add a line into our spreadsheet to discuss further.

tbg replied:
Arguably since there is a truncation happening here we should be "staging" the truncation like we do for regular truncations: updated the truncated state first (which would necessarily lead to an empty log - we should assert against the case in which that is not true), then clear the cache, then perform the disk updates. WDYT?

pav-kv replied:
Good point, there should be something similar to `stagePendingTruncationRaftMuLocked` before the write. Also, we should clear sideloaded storage after write/sync, like in `finalizeTruncationRaftMuLocked`. There is a TODO [here](https://docs.google.com/document/d/1ugGKYszAow71PSK7PeN3DrVYZiUCsE7mKCY-8QLDTSY/edit?disco=AAABZixpfaU).

tbg replied:
In an ideal world (and maybe a world that's just a PR away), should we be calling the _actual_ `stagePendingTruncationRaftMuLocked` (and `finalizeTruncationRaftMuLocked` after)?

pav-kv replied:
Yeah, I think it should just work. There are 2 kinds of methods on the cache and sideloaded storage though: `TruncateTo` and `Clear`.

The `stage` and `finalize` method does the former, whereas we would kind want the latter here. However, `TruncateTo(snap.index)` is functionally equivalent.

tbg committed:
kvserver: leave a comment/TODO about entry cache

tbg committed:
kvserver: rearrange struct fields

tbg committed:
kvserver: rename a struct field

tbg committed:
kvserver: rename for clarity

There are various stages of snapshots. Rename the types and method so that it's
clear that we're preparing a snapshot for application (vs. for sending, etc).

tbg committed:
kvserver: rename a field

tbg committed:
kvserver: add comment

tbg committed:
kvserver: remove prepareSnapApply return type

tbg committed:
kvserver: call subsumption code unconditionally

It works (or should work, based on my reading) when there are no subsumed descs
and in that case is a no-op.

tbg committed:
kvserver: fix comments referring to preDestroyRaftMuLocked

tbg committed:
kvserver: rename ingest -> applyAsIngest

tbg commented:
TFTR! I added a number of commits to address your suggestions and some additional improvements, PTAL.

pav-kv commented on pkg/kv/kvserver/snapshot_apply_prepare.go:
```diff
@@ -51,15 +55,12 @@ func prepareSnapApply(
-			return roachpb.Span{}, nil, err
-		}
-		clearedSubsumedSpans = spans
+	clearedSubsumedSpans, err := clearSubsumedReplicaDiskData(
```
I timed out on a couple of minutes looking at `clearSubsumedReplicaDiskData` trying to understand why it's a no-op if `subsumedDescs` is empty. It probably indicates that it's better to make it explicit (in its comment and/or have an `if` in its body).

Particularly, why is this second loop in it a no-op?
https://github.com/cockroachdb/cockroach/blob/df7d3153243b3b680cdc2c7b8474bbd6f11ed9e9/pkg/kv/kvserver/replica_raftstorage.go#L930

We used to call this unconditionally though (so the change is good), is the `if len > 0` a recent addition? https://github.com/cockroachdb/cockroach/blob/df7d3153243b3b680cdc2c7b8474bbd6f11ed9e9/pkg/kv/kvserver/replica_raftstorage.go#L553-L564.

pav-kv replied:
On a second attempt to read it, I see why it's no-op: it compares keys and comparisons never succeed. I still think we should make the zero case explicit to remove doubt, and also seems like there is wasted work there generating and comparing keys.

tbg replied:
> is the if len > 0 a recent addition?

Yes, this is unfortunate AI jank - it was introduced in "kvserver: move subsumed replica disk data clearing into prepareSnapshot" and I didn't catch it during self-review.

I'll add a comment about the case in which subsumedDescs is nil. I'm mildly -1 on bailing early in this method: if we don't trust this method to do the right thing if nothing got subsumed, how can we believe that it does the right thing in general?

pav-kv replied:
> how can we believe that it does the right thing in general?

By testing it :)

The argument is not in the lack of trust in the code, but it's in readability (which is only a proxy for trust). It took me maybe 5 minutes to figure out how the empty case behaves and spot that it's a no-op.

pav-kv replied:
LGTM, thanks. Mind popping this somewhere (e.g. in the `clearSubsumedReplicaDiskData` comment)?

```diff
diff --git a/pkg/kv/kvserver/replica_raftstorage.go b/pkg/kv/kvserver/replica_raftstorage.go
index e19716a5995..f1447d06094 100644
--- a/pkg/kv/kvserver/replica_raftstorage.go
+++ b/pkg/kv/kvserver/replica_raftstorage.go
@@ -849,6 +849,8 @@ func rewriteRaftState(
 // method requires that each of the subsumed replicas raftMu is held, and that
 // the Reader reflects the latest I/O each of the subsumed replicas has done
 // (i.e. Reader was instantiated after all raftMu were acquired).
+//
+// NB: does nothing if subsumedDescs is empty.
 func clearSubsumedReplicaDiskData(
        ctx context.Context,
        st *cluster.Settings,
```

pav-kv commented on pkg/kv/kvserver/store.go:
```diff
@@ -752,7 +752,7 @@ prose below, and the source .dot file is in store_doc_replica_lifecycle.dot.
 	          +------------------ |       Absent        | ---------------------------------------------------------------------------------------------------+
 	          |                   +---------------------+                                                                                                    |
-	          |                     |                        Subsume              Crash          applySnapshot                                               |
+	          |                     |                        Subsume              Crash          applySnapshotRaftMuLocked                                               |
```
Fix padding in the tail?

tbg committed:
kvserver: move an assertion

tbg committed:
kvserver: simplify clearSubsumedReplicaDiskData

It may be a little easier on the eyes (and brain) now.

tbg commented:
TFTR! I'll send a follow-up about the truncation shortly.

bors r+
